crates/ruma-common/src/events/audio.rs

```diff
@@ -0,0 +1,180 @@
+    pub waveform: Waveform,
+}
+
+impl AudioContent {
+    /// Creates a new empty `AudioContent`.
+    pub fn new() -> Self {
+        Self::default()
+    }
+}
+
+/// The waveform representation of audio content.
+///
+/// Must include between 30 and 120 `Amplitude`.
+///
+/// To build this, use the `TryFrom` implementations.
+#[derive(Clone, Debug, Default, Deserialize, Serialize)]
+#[serde(try_from = "WaveformSerDeHelper")]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+pub struct Waveform(Vec<Amplitude>);
+
+impl Waveform {
+    /// The smallest number of values contained in a `Waveform`.
+    pub const MIN_LENGTH: usize = 30;
+
+    /// The largest number of values contained in a `Waveform`.
+    pub const MAX_LENGTH: usize = 120;
+
+    /// The amplitudes of this `Waveform`.
+    pub fn amplitudes(&self) -> &[Amplitude] {
+        &self.0
+    }
+
+    /// Whether this `Waveform` is empty.
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+}
+
+/// An error encountered when trying to convert to a `Waveform`.
+#[derive(Copy, Clone, Debug, Eq, Hash, PartialEq, thiserror::Error)]
+#[non_exhaustive]
+pub enum WaveformError {
+    /// There are more than [`Waveform::MAX`] values.
+    #[error("too many values")]
+    TooManyValues,
+    /// There are less that [`Waveform::MIN`] values.
+    #[error("not enough values")]
+    NotEnoughValues,
+}
+
+impl TryFrom<Vec<Amplitude>> for Waveform {
+    type Error = WaveformError;
+
+    fn try_from(value: Vec<Amplitude>) -> Result<Self, Self::Error> {
+        if value.len() < Self::MIN_LENGTH {
+            Err(WaveformError::NotEnoughValues)
+        } else if value.len() > Self::MAX_LENGTH {
+            Err(WaveformError::TooManyValues)
+        } else {
+            Ok(Self(value))
+        }
+    }
+}
+
+impl TryFrom<&[Amplitude]> for Waveform {
+    type Error = WaveformError;
+
+    fn try_from(value: &[Amplitude]) -> Result<Self, Self::Error> {
+        Self::try_from(value.to_owned())
+    }
+}
+
+/// The amplitude of a waveform sample.
+///
+/// Must be an integer between 0 and 1024.
+#[derive(Clone, Copy, Debug, Default, Hash, PartialEq, Eq, PartialOrd, Ord, Serialize)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+pub struct Amplitude(UInt);
+
+impl Amplitude {
+    /// The smallest value that can be represented by this type, 0.
+    pub const MIN: u16 = 0;
+
+    /// The largest value that can be represented by this type, 1024.
+    pub const MAX: u16 = 1024;
+
+    /// Creates a new `Amplitude` with the given value.
+    ///
+    /// It will saturate if it is bigger than [`Amplitude::MAX`].
+    pub fn new(value: u16) -> Self {
+        value.into()
+    }
+
+    /// The value of this `Amplitude`.
+    pub fn value(&self) -> UInt {
+        self.0
+    }
+}
+
+impl From<u8> for Amplitude {
+    fn from(value: u8) -> Self {
+        Self(value.into())
+    }
+}
```

Comment:
I don't think we should have this impl. If sbd. is representing the amplitude as a `u8`, they're doing things wrong.